[tricixg] iP

tricixg opened the pull request:
### Broccoli 

> Broccoli is a smart dinosaur

- text-based
- easy to learn
- FAST SUPER FAST to use
- All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

`System.out.println(" <|°▿▿▿▿°|/ ")`

And it is **_FREE_**!

Features:
- [ ] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

Commit:
Add Gradle support

tricixg committed:
Level-1 Greet, Echo, Exit

- Customised welcome message
- gets user input
- echo user input
- exits when user inputs "bye"

tricixg committed:
Level-2:Add, List

-add items to the list
-display items in the list

tricixg committed:
Level-3: Mark as Done

-ability to mark and unmark tasks
-Task class created

tricixg committed:
Task:Deadline implemented

tricixg committed:
Level 4. ToDos, Events, Deadlines

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

tricixg committed:
Level-5: Handle Errors

- handle errors
- cleaned up code by creating new file for each class

tricixg committed:
Level-6: Delete

- deletes task
- declared divider as variable

tricixg committed:
A-TextUiTesting

tricixg committed:
Level-7

Implement save function
able to load task from previous sessions

tricixg committed:
Level 8: DateTime added

tricixg committed:
Merge pull request #1 from tricixg/branch-Level-7

Level-7

tricixg committed:
Merge branch 'master' into branch-Level-8

tricixg committed:
Merge pull request #2 from tricixg/branch-Level-8

Level 8: DateTime added

tricixg committed:
Level-8

BoAi01 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,114 @@
+import java.time.format.DateTimeParseException;
+
+import java.util.ArrayList;
+
+
```
Based on conventions [here](https://se-education.org/guides/conventions/java/intermediate.html), it might be better to remove the extra empty lines here (perhaps only one empty line is needed). : )

BoAi01 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,114 @@
+
+
+import util.DukeException;
+
+public class Deadline extends Task {
+
```
I like the implementation here, but it would be better if we have a Javadoc for the class as well as its methods. :D

BoAi01 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+     
+        //print welcome message, ask for user input
+        System.out.println("  ───── ･ ｡ﾟ☆: *.☽ .* :☆ﾟ. ───── \n Hello! I'm Broccoli the dinosaur \n           <|°▿▿▿▿°|/ \n      What can I do for you? \n   ──── ･ ｡ﾟ☆: *.☽ .* :☆ﾟ. ────" ); 
+        //getInput();
+        
+
+
```
Perhaps we can remove some empty lines and used codes to make the program look neater. :grinning:

BoAi01 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,114 @@
+
+            if(isDate(date)){
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+                LocalDate ld = LocalDate.parse(date, formatter);
+                Deadline d = new Deadline(desc, ld);
```
I am not sure, but I guess it looks better to readers if the variable names are more descriptive, e.g., `localDate` and `deadline`, so that they understand at first glance.

BoAi01 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+
+
+     
+        //print welcome message, ask for user input
+        System.out.println("  ───── ･ ｡ﾟ☆: *.☽ .* :☆ﾟ. ───── \n Hello! I'm Broccoli the dinosaur \n           <|°▿▿▿▿°|/ \n      What can I do for you? \n   ──── ･ ｡ﾟ☆: *.☽ .* :☆ﾟ. ────" ); 
+        //getInput();
```
I like the interesting name of the robot!

BoAi01 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,173 @@
+                System.out.println(divider);
+            }
+        } else {
+            // for(int j=1; j< splitInput.length; j++){
+            //     if(splitInput[j].equals("/from")){
+            //         for (int k=1; k< j-1; k++){
+            //             splitInput[1] = splitInput[1] + " " + splitInput[k+1];
+            //         }
+            //         for(int i = j + 1; i < splitInput.length; i++) {
+            //             splitInput[i] = splitInput[i].replace("/to", "to:");
+            //         }
+            //         for (int l=splitInput.length-1; l > j +1; l--){
+            //             splitInput[splitInput.length-1] = splitInput[l-1]+" "+splitInput[splitInput.length-1];
+            //         }
+            //     } else {
+            //         splitInput[j] = splitInput[j];
+            //     }
+            // }
+               
+            // String duration = splitInput[splitInput.length-1];
+            // String desc = splitInput[1];
+           
```
Perhaps these lines can be removed once you feel ready finalizing the program?

BoAi01 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,114 @@
+            System.out.println(isDate(date));
+
+
+            if(isDate(date)){
```
If I remember correctly, we had better add some space outside parentheses, i.e., 
```
if (isDate(date)) {
```
By the way, the naming of `isDate` sounds good!

BoAi01 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+        
+        try {
+            SaveData.loadFile(array);
+        } catch(Exception e) {
+            System.out.println(e.getMessage());
+        }
```
Should we be more specific about what exception to catch? I guess you want the following
```
} catch(IOException e) {
```
But I am not very sure about the exact name of the exception that `loadFile` could throw, so please feel free to correct me if I am wrong. : )

BoAi01 reviewed (commented):
Overall, this is a good effort. The code is easily understandable, but it could be better if there are more Javadoc and fewer lines of commented-out codes. The naming for variables, classes, and methods is generally fine.

EvitanRelta commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,173 @@
+            String event = combinedString.split(" ", 2)[1];
+            String[] event_Arr = event.split(" /from", 2);
+
+                if (event_Arr.length == 2) {
+                    String desc = event_Arr[0];
+                    String[] period_Arr = event_Arr[1].split(" /to");
+                    if (period_Arr.length == 2) {
+
+                        String from = period_Arr[0].strip();
+                        String to = period_Arr[1].strip();;
+
+                        if(isDate(from) && isDate(to)){
+                            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+                            LocalDate ldFrom = LocalDate.parse(from, formatter);
+                            LocalDate ldTo = LocalDate.parse(to, formatter);
+                            Event e = new Event(desc, ldFrom, ldTo);
+                            array.add(e);
+                            System.out.println(divider);
+                            System.out.println("     Got it. I've added this task:");
+                            System.out.println("     " + e.toString());
+                            System.out.println("     Now you have " + array.size() + " tasks in the list.");
+                            System.out.println(divider);
+                        } else {
+                            Event e = new Event(desc, from, to);
+                            array.add(e);
+                            System.out.println(divider);
+                            System.out.println("     Got it. I've added this task:");
+                            System.out.println("     " + e.toString());
+                            System.out.println("     Now you have " + array.size() + " tasks in the list.");
+                            System.out.println(divider);
+                        }
+
+                       
+                    }
+                }
```
Perhaps, this make use of [guard clauses] to reduce the indentations of the `if` statements, as per the ["make-the-happy-path-prominent"][link] and ["Avoid Deep Nesting"][nesting] coding guidelines, like so:





[guard clauses]: http://wiki.c2.com/?GuardClause
[link]: https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent
[nesting]: https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting

```java
if (event_Arr.length != 2) {
    return;
}

String desc = event_Arr[0];
String[] period_Arr = event_Arr[1].split(" /to");
if (period_Arr.length != 2) {
    return;
}

String from = period_Arr[0].strip();
String to = period_Arr[1].strip();;

if(isDate(from) && isDate(to)){
    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    LocalDate ldFrom = LocalDate.parse(from, formatter);
    LocalDate ldTo = LocalDate.parse(to, formatter);
    Event e = new Event(desc, ldFrom, ldTo);
    array.add(e);
    System.out.println(divider);
    System.out.println("     Got it. I've added this task:");
    System.out.println("     " + e.toString());
    System.out.println("     Now you have " + array.size() + " tasks in the list.");
    System.out.println(divider);
} else {
    Event e = new Event(desc, from, to);
    array.add(e);
    System.out.println(divider);
    System.out.println("     Got it. I've added this task:");
    System.out.println("     " + e.toString());
    System.out.println("     Now you have " + array.size() + " tasks in the list.");
    System.out.println(divider);
}
```

EvitanRelta commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,114 @@
+
+    public static void createDeadlineTask(ArrayList<Task> array, String[] splitInput) {
+        if (splitInput.length == 1 || splitInput[1].equals("")){
+            try {
+                throw new DukeException("deadline");
+            } catch (Exception e) {
+                System.out.println(divider);
+                System.out.println(e.toString());
+                System.out.println(divider);
+            }
```
I'm not too sure what this try-catch is trying to do though.

I'm assuming that you're expecting the throwing of a `DukeException` might fail?

EvitanRelta commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+                break;
+            }
+            //Displays the list of items 
+            else if (combined.equals("list")){
+                Task.displayList(array);
+                
+            } 
+            //Mark the task as done
+            else if(splitInput[0].equals("mark")){
+                Task.markTask(array, splitInput);
+            } 
+            //Unmark the task as done
+            else if(splitInput[0].equals("unmark")){
+                Task.unmarkTask(array, splitInput);
+            }
+            //Creates a Deadline type Task 
+            else if(splitInput[0].equals("deadline")){
+                Deadline.createDeadlineTask(array, splitInput);
+            } 
+            //Creates an Event type Task
+            else if(splitInput[0].equals("event")) {
+                Event.createEventTask(array, splitInput);
+            } 
+            //Creates a Todo type Task
+            else if(splitInput[0].equals("todo")) {
+                Todo.createTodoTask(array, splitInput);
+            }
+            //Deletes a Task 
+            else if(splitInput[0].equals("delete")) {
+                Task.deleteTask(array, splitInput);
+            } else {
```
I like how each `if` block has the same level of abstraction. It follows the [SLAP] coding guideline well! :D

[slap]: https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#slap-hard

EvitanRelta commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+            else if(splitInput[0].equals("delete")) {
+                Task.deleteTask(array, splitInput);
+            } else {
+                try {
+                    throw new IllegalArgumentException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+                } catch (Exception e) {
+                    System.out.println(divider);
+                    System.out.println(e.getMessage());
+                    System.out.println(divider);
+                }
```
Is there a specific reason to do a `try` on the `throw new IllegalArgumentException`?

I'm assuming you're trying to catch any errors thrown by the `Task.displayList`, `Task.deleteTask`, etc. methods? If so, perhaps you'll need to put those methods in the `try` block to catch those errors.

EvitanRelta commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,79 @@
+        System.out.println(divider);
+    }
+
+    public static void unmarkTask(ArrayList<Task> array, String[] splitInput) {
+        array.get((Integer.parseInt(splitInput[1])-1)).markAsUnDone();
+        System.out.println(divider);
+        System.out.println("      OK, I've marked this task as not done yet:");
+        System.out.println("      "+ array.get((Integer.parseInt(splitInput[1])-1)).toString());
+        System.out.println(divider);
+    }
+
+    public static void markTask(ArrayList<Task> array, String[] splitInput) {
+        array.get((Integer.parseInt(splitInput[1])-1)).markAsDone();
+        System.out.println(divider);
+        System.out.println("      Nice! I've marked this task as done:");
+        System.out.println("      "+ array.get((Integer.parseInt(splitInput[1])-1)).toString());
+        System.out.println(divider);
+    }
+
+    public static void deleteTask(ArrayList<Task> array, String[] splitInput) {
+        System.out.println(divider);
+        System.out.println("      Noted. I've removed this task:");
+        System.out.println("      "+ array.get((Integer.parseInt(splitInput[1])-1)).toString());
+        array.remove((Integer.parseInt(splitInput[1])-1));
+        System.out.println("      Now you have " + array.size() + " tasks in the list.");
+        System.out.println(divider);
+    }
```
There's a lot of repeated `System.out.println` calls in the code, all of which adds a divider on the top and bottom, and indents the text. 

To [minimize code duplication][dry], perhaps you can abstract out them out to a `say` or `printString` method to print a multi-line string along with the divider and indentation? Like:

```java
public static void say(String whatToSay) {
    int indentLevel = 6
    String indentation = " ".repeat(indentLevel);
    String indentedInput = whatToSay.replaceAll("(?<=^|\n)", indentation);
    
    System.out.println(divider);
    System.out.println(indentedInput);
    System.out.println(divider);
}
```

And then call the method like:

```java
public static void markTask(ArrayList<Task> array, String[] splitInput) {
    array.get((Integer.parseInt(splitInput[1])-1)).markAsDone();
    say("Nice! I've marked this task as done:\n"
            + array.get((Integer.parseInt(splitInput[1])-1)).toString());
}
```

[dry]: https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#minimize-code-duplication

tricixg committed:
A-MoreOOP

added Parser, Storage and UI

tricixg committed:
A-Packages

added previously

tricixg committed:
Merge pull request #4 from tricixg/add-gradle-support

Add Gradle support

tricixg committed:
A-JUnit

added tests for deadline and todo

tricixg committed:
A-JavaDoc

added docs for classes
Deadline
Event
Task
Todo

tricixg committed:
Level-9

Give users a way to find a task by searching for a keyword.
find book

     Here are the matching tasks in your list:
     1.[T][X] read book
     2.[D][X] return book (by: June 6th)

tricixg committed:
revert javadoc

tricixg committed:
Update TaskList

tricixg committed:
remove javadoc

tricixg committed:
add javadoc

tricixg committed:
remove javadoc

tricixg committed:
add javadocs

tricixg committed:
Merge pull request #5 from tricixg/Fix-detached-head

Fix detached head

tricixg committed:
A-JavaDoc

tricixg committed:
Merge branch 'master' of https://github.com/tricixg/ip

tricixg committed:
style

tricixg committed:
A-CheckStyle

tricixg committed:
Merge pull request #6 from tricixg/branch-A-CheckStyle

A-CheckStyle

tricixg committed:
Added GUI

tricixg committed:
Level-10

GUI for duke bot
customised to broccoli
cleaned up code

tricixg committed:
Merge pull request #7 from tricixg/branch-Level-10

Branch level 10

tricixg committed:
Create MANIFEST.MF

tricixg committed:
A-Assertions

Assert feature added in Storage class.

This is to ensure that the file path is not empty when loading data

tricixg committed:
Merge pull request #8 from tricixg/branch-A-Assertions

A-Assertions

tricixg committed:
A-CodeQuality

Improved Code Quality following the java coding standards

Added JavaDocs to all methods.

Corrected the style of code using CheckStyle

tricixg committed:
Merge pull request #9 from tricixg/branch-A-CodeQuality

A-CodeQuality

tricixg committed:
C-Help

New users may not know what to do when first launched.

Added command 'help' which provides users with a list of commands.

Instruction and greeting also shown when program is launched.

tricixg committed:
A-BetterGUI

customised the dialogbox colour

tricixg committed:
A-MoreTesting

Added more test for Deadline, Event and Todo class

tests include creation, completion and toString

tricixg committed:
Add UI.png

tricixg committed:
UPDATE USERGUIDE

tricixg committed:
Update README.md

tricixg committed:
Ignores case

ignores case for input.
command HELP, Help can be accepted

tricixg committed:
bug fix

fixed bug of deleting task

added error message when invalid task number is input when marking, unmarking and deleting tasks

tricixg committed:
Update README.md

tricixg committed:
Update README.md

tricixg committed:
bug fix: spaces

used .trim() to remove any leading or trailing white spaces in users input which may cause an error.

tricixg committed:
edit build.gradle

tricixg committed:
update README

tricixg committed:
improve code quality
